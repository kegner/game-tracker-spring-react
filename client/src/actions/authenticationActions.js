export const SET_USER = "SET_USER";
export const CLEAR_USER = "CLEAR_USER";

export const setUserData = (user) => ({
  type: SET_USER,
  payload: { user }
});

export const clearUserData = () => ({
    type: CLEAR_USER
});