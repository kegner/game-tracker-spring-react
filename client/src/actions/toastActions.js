export const SHOW_SUCCESS = "SHOW_SUCCESS";
export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_TOAST = "HIDE_TOAST";

export const showSuccessToast = (message) => ({
  type: SHOW_SUCCESS,
  payload: { message }
});

export const showErrorToast = (message) => ({
  type: SHOW_ERROR,
  payload: { message }
});

export const hideToast = (id) => ({
  type: HIDE_TOAST,
  payload: { id }
});