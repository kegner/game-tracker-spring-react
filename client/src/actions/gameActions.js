import axios from 'axios';
import { showSuccessToast, showErrorToast } from './toastActions';

export const SELECT_GAME = "SELECT_GAME";
export const FETCH_GAMES = "FETCH_GAMES";
export const CREATE_GAME = "CREATE_GAME";
export const UPDATE_GAME = "UPDATE_GAME";
export const DELETE_GAME = "DELETE_GAME";
export const START_LOADING = "START_LOADING";
export const START_SAVING = "START_SAVING";
export const STOP_LOADING = "STOP_LOADING";
export const STOP_SAVING = "STOP_SAVING";

const fetchSuccess = (games) => ({
  type: FETCH_GAMES,
  payload: { games }
});

const createSuccess = (game) => ({
  type: CREATE_GAME,
  payload: { game }
});

const updateSuccess = (game) => ({
  type: UPDATE_GAME,
  payload: { game }
});

const deleteSuccess = (id) => ({
  type: DELETE_GAME,
  payload: { id }
});

const startLoading = () => ({
  type: START_LOADING
});

const startSaving = () => ({
  type: START_SAVING
});

const stopLoading = () => ({
  type: STOP_LOADING
});

const stopSaving = () => ({
  type: STOP_SAVING
});

const errorHandler = (error, dispatch) => {
  const validResponses = [404, 409];
  if (error && error.response && validResponses.includes(error.response.status)) {
    // this is the expected error format for Spring exception messages
    dispatch(showErrorToast(error.response.data.message));
  } else {
    console.error(error);
  }
};

export const selectGame = (game) => ({
  type: SELECT_GAME,
  payload: { game }
});

export const fetchGames = (userId) => async (dispatch) => {
  dispatch(startLoading());

  try {
    const res = await axios.get("/api/v1/games", { params: { userId } });
    dispatch(fetchSuccess(res.data.games));
  } catch (error) {
    dispatch(stopLoading());
    errorHandler(error, dispatch);
  }
};

export const createGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.post("/api/v1/games", game);

    dispatch(createSuccess(res.data));
    dispatch(showSuccessToast("Game added."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const updateGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.put("/api/v1/games", game);

    dispatch(updateSuccess(res.data));
    dispatch(showSuccessToast("Game updated."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const deleteGame = (id) => async (dispatch) => {
  dispatch(startSaving());

  try {
    await axios.delete(`/api/v1/games/${id}`);
    dispatch(deleteSuccess(id));
    dispatch(showSuccessToast("Game deleted."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};