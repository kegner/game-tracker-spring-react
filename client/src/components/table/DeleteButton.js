import React from 'react'
import { useDispatch } from 'react-redux';
import { deleteGame } from '../../actions/gameActions';

const DeleteButton = (props) => {

  const { id } = props;

  const dispatch = useDispatch();

  const deleteHandler = (e) => {
    e.stopPropagation();
    dispatch(deleteGame(id));
  }

  return (
    <button className="btn btn-danger" aria-label="Delete" onClick={deleteHandler}>
      <i className="bi bi-trash"></i>
    </button>
  )
}

export default DeleteButton
