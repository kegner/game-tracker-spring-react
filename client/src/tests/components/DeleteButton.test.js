import React from 'react';
import { Provider } from "react-redux";
import { render, fireEvent } from '@testing-library/react';
import { START_SAVING, DELETE_GAME } from '../../actions/gameActions';
import { SHOW_SUCCESS } from '../../actions/toastActions';
import { createTestStore } from '../store/testStore';
import DeleteButton from '../../components/table/DeleteButton';

let store;
let dispatch;
let component;
let container;
let asFragment;

jest.mock("axios");

describe("DeleteButton", () => {

  beforeEach(() => {
    store = createTestStore({
      mockStore: true
    });
    dispatch = store.dispatch;
    component = render(
      <Provider store={store}>
        <DeleteButton id="123" />
      </Provider>
    );
    container = component.container;
    asFragment = component.asFragment;
  });

  it("exists", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("deletes game", async () => {
    const expectedActions = [
      { type: START_SAVING },
      { type: DELETE_GAME, payload: { id: "123" } },
      { type: SHOW_SUCCESS, payload: { message: "Game deleted." } }
    ];

    const button = container.querySelector("button");
    expect(dispatch).not.toHaveBeenCalled();
    fireEvent.click(button);
    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toEqual(expectedActions);
  });

});