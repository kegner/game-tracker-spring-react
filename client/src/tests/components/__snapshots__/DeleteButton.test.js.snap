// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteButton exists 1`] = `
<DocumentFragment>
  <button
    aria-label="Delete"
    class="btn btn-danger"
  >
    <i
      class="bi bi-trash"
    />
  </button>
</DocumentFragment>
`;
